### PR TITLE
[onert] Introduce DepthwiseConvOp in cpu kernel

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit introduces DepthwiseConvOp in cpu kernel.
This kernel uses depthwise_conv and bias_op eigen functions.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

### CODE REVIEW

**Comment by hseok-oh** [comment]
```
In file included from /home/runner/work/ONE/ONE/runtime/onert/backend/cpu/ops/DepthwiseConvolutionLayer.cc:20:
In file included from /home/runner/work/ONE/ONE/compute/cker/include/cker/operation/DepthwiseConv.h:32:
/home/runner/work/ONE/ONE/compute/cker/include/cker/eigen/bias_op.h:100:7: error: unused variable 'channel_dim' [-Werror,-Wunused-variable]
  int channel_dim = input_shape.DimensionsCount() - 1;
      ^
1 error generated.
gmake[3]: *** [runtime/onert/backend/cpu/CMakeFiles/onert_backend_cpu.dir/build.make:272: runtime/onert/backend/cpu/CMakeFiles/onert_backend_cpu.dir/ops/DepthwiseConvolutionLayer.cc.o] Error 1
```

**Comment by jyoungyun** [comment]
Need #13582

File: compute/cker/include/cker/operation/DepthwiseConv.h [hunk 1]
```diff
@@ -180,6 +182,47 @@ inline void DepthwiseConv(const DepthwiseConvParams &params, const Shape &input_
...
+  const int pad_width = params.padding_values.width;
+  const T activation_min = params.float_activation_min;
+  const T activation_max = params.float_activation_max;
+
+  depthwise_conv_op::LaunchDepthwiseConvOp<Eigen::ThreadPoolDevice, T>()(
+    batch, input_height, input_width, input_depth, filter_height, filter_width, depth_multiplier,
+    stride, pad_height, pad_width, output_height, output_width, output_depth, input_data,
+    filter_data, padded_filter_data, pad_filter, filter_buffers_data, output_data);
+
+  if (bias_data != nullptr)
```
Comment (ys44kim) [inline compute/cker/include/cker/operation/DepthwiseConv.h:219]: i think it would be better to add a exception action for the case where bias_data is nullptr.
for example, add assert (bias_data) as follows refer to biasHelper,
```
template <typename T>
void biasHelper(const Shape &bias_shape, const T *bias_data, const Shape &input_shape,
                T *input_data, T activation_min, T activation_max)
{
  int channel_dim = input_shape.DimensionsCount() - 1;

  assert(input_shape.Dims(channel_dim) == bias_shape.Dims(0));
  assert(input_data);
  assert(bias_data);
```

File: compute/cker/include/cker/operation/DepthwiseConv.h [hunk 2]
```diff
@@ -180,6 +182,47 @@ inline void DepthwiseConv(const DepthwiseConvParams &params, const Shape &input_
...
+  const int filter_height = filter_shape.Dims(1);
+  const int filter_width = filter_shape.Dims(2);
+  const int output_height = output_shape.Dims(1);
+  const int output_width = output_shape.Dims(2);
+  const int stride = params.stride_height;
+  const int depth_multiplier = params.depth_multiplier;
+  const int pad_height = params.padding_values.height;
+  const int pad_width = params.padding_values.width;
+  const T activation_min = params.float_activation_min;
+  const T activation_max = params.float_activation_max;
```
Comment (ragmani) [inline compute/cker/include/cker/operation/DepthwiseConv.h:212]: I'm not sure if non-float types use `params.float_activation_min` and we have to support those types for DepthwiseConv op. So, I think it's better to change this function from function template to normal function that supports only `float`.

File: compute/cker/include/cker/operation/DepthwiseConv.h (hunk 1 above)
Comment (jyoungyun) [reply compute/cker/include/cker/operation/DepthwiseConv.h:219]: Does the circle model always have `bias_data` in DepthwiseConv2D operation? The TensorFlow API supports the `use_bias` attribute for null bias data. I am not sure that `bias_data` is not always nullptr.

Comment (ragmani) [reply compute/cker/include/cker/operation/DepthwiseConv.h:218]: @jyoungyun You are right. Tensorflow API supports optional bias. But onert does not support optional bias for DepthwiseConv2D op yet. So, if a circle model has it, that model won't run on onert.
Anyway, I think it's better to a test that does not have bias data if you'd like to support the optional bias on cker.

File: compute/cker/src/DepthwiseConv.test.cc [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    nnfw::cker::Shape filter_shape{1, 2, 2, 1}; // 1, h, w, c
+    std::vector<float> filter = {1.0, 2.0, 3.0, 4.0};
+    nnfw::cker::Shape bias_shape{1};
+    nnfw::cker::Shape output_shape{1, 3, 3, 1}; // n, h, w, c
+    std::vector<float> expected = {4.0, 0.0, 3.0, 0.0, 0.0, 0.0, 2.0, 0.0, 1.0};
+
+    DepthwiseConvVerifier<float> verifier;
+    verifier.prepare(output_shape, filter_shape);
+    verifier.checkDeath(params, input_shape, input.data(), filter_shape, filter.data(), bias_shape,
+                        nullptr, output_shape, expected.data());
```
Comment (jyoungyun) [inline compute/cker/src/DepthwiseConv.test.cc:318]: The DepthwiseConv2D operation always has a `bias_data`, so I removed the `if(bias_data != nullptr)` check case. Instead, I add a new test case to check if app makes an assertion if the bias_data is null.

File: compute/cker/src/DepthwiseConv.test.cc [hunk 4]
```diff
@@ -84,6 +85,20 @@ template <typename T> class DepthwiseConvVerifier
...
   }
 
+  void checkDeath(const nnfw::cker::DepthwiseConvParams &params,
+                  const nnfw::cker::Shape &input_shape, const T *input_data,
+                  const nnfw::cker::Shape &filter_shape, const T *filter_data,
+                  const nnfw::cker::Shape &bias_shape, const T *bias_data,
+                  const nnfw::cker::Shape &output_shape, const T *expected)
+  {
+    std::vector<T> output(output_shape.FlatSize());
+    EXPECT_DEATH(nnfw::cker::DepthwiseConvOp(params, input_shape, input_data, filter_shape,
```
Comment (ragmani) [inline compute/cker/src/DepthwiseConv.test.cc:95]: Binaries built in release mode do not execute assertions. You need to ensure this line is executed only on binaries built in debug mode.

File: compute/cker/src/DepthwiseConv.test.cc (hunk 3 above)
Comment (ragmani) [reply compute/cker/src/DepthwiseConv.test.cc:318]: What I meant was that it would be better to add a test to check if optional bias is supported. But I think it's not bad to add this test instead of my intention.

File: compute/cker/src/DepthwiseConv.test.cc [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+                                             filter_data, bias_shape, bias_data,
+                                             _padded_filter.data(), _use_padded_filter,
+                                             _filter_buffers.data(), output_shape, output.data()),
+                 "");
```
Comment (ragmani) [inline compute/cker/src/DepthwiseConv.test.cc:99]: This regular expression(or matcher) probably won't fail at all. That means this test is always passed even if the condition of the corresponding assertion is not `true`.

File: compute/cker/src/DepthwiseConv.test.cc (hunk 3 above)
Comment (ys44kim) [reply compute/cker/src/DepthwiseConv.test.cc:318]: there is no code that access bias_data, and it is passed as argument , there is no need to check for nullptr. 
of course, even when there is no explicit nullptr input, sometimes the static analysis tool judges it to be an error when accessing the variable without checking for nullptr. of course, this does not apply in this case.
so, i think it is a good choice.

File: compute/cker/src/DepthwiseConv.test.cc (hunk 4 above)
Comment (jyoungyun) [reply compute/cker/src/DepthwiseConv.test.cc:95]: `EXPECT_DEBUG_DEATH` executes statement if it is RELEASE mode. It makes an systeam halt because the bias_data is null. How about wrapping it in define so that it only runs in debug mode?

File: compute/cker/src/DepthwiseConv.test.cc (hunk 3 above)
Comment (jyoungyun) [reply compute/cker/src/DepthwiseConv.test.cc:318]: I'm sorry but I couldn't catch your intention. :(

> add a test to check if optional bias is supported

Could you explain more about this test case?

Comment (ragmani) [reply compute/cker/src/DepthwiseConv.test.cc:318]: > Could you explain more about this test case?

`biasHelper` abort if `bias_data` is nullptr. It means that optional bias(`nullptr`) is not allowed for DepthwiseConv2D.
But what I meant was that optional bias is allowed and the test verifies whether the DepthwiseConv2D kernel works well with optional bias(`nullptr`).

File: compute/cker/src/DepthwiseConv.test.cc (hunk 4 above)
Comment (ragmani) [reply compute/cker/src/DepthwiseConv.test.cc]: > EXPECT_DEBUG_DEATH executes statement if it is RELEASE mode. 

You are right. `EXPECT_DEBUG_DEATH` executes statement if it is RELEASE mode. 

> It makes an systeam halt because the bias_data is null.

I understand this now. If `bias_data` is `nullptr`, system is halted at another location, not `assert` statement in release mode.

> How about wrapping it in define so that it only runs in debug mode?

As we talked about offline, it would be better to allow optional bias and change that test to verify `DepthwiseConv2D` kernel with optional bias.

File: compute/cker/src/DepthwiseConv.test.cc (hunk 5 above)
Comment (ragmani) [reply compute/cker/src/DepthwiseConv.test.cc]: As we saw offline, `""` means to allow all error messages, not to prevent the test from failing. The reason this test passed in release mode was because the test died for another reason(maybe seg fault), not the assertion.

**Comment by jyoungyun** [comment]
@ys44kim 

> there is no code that access bias_data, and it is passed as argument , there is no need to check for nullptr.
> of course, even when there is no explicit nullptr input, sometimes the static analysis tool judges it to be an error when accessing the variable without checking for nullptr. of course, this does not apply in this case.
> so, i think it is a good choice.

The code was changed to support optional bias data in DepthwiseConv2D cpu kernel. :)

> i think it would be better to add a exception action for the case where bias_data is nullptr.
> for example, add assert (bias_data) as follows refer to biasHelper,

In ONERT, the DepthwiseConv2D always has bias data. However, this behaviour can be changed because the DepthwiseConv2D does not necessarily require bias data.
So, I want to support `optional bias` in DepthwiseConv2D cpu kernel.
I added a test case for `optional bias` test case for verifying this cpu kernel.

Please take another look. :)

**Comment by ys44kim** [comment]
@jyoungyun 
please understand that this review may be incorrect as I do not know the exact domain knowledge and onert.

and, In onert the DepthwiseConv2D always has bias data  (please tell me if this is wrong.)
so i understood that only those with bias data would be considered. i am sorry that it is my misunderstanding.

and, In fact, I don't understand why it is tested when bias is nullptr because onert does not support optional bias.
>  onert does not support optional bias for DepthwiseConv2D op yet.   

of course, If onert code was added to support optional bias (maybe I couldn't find it) and if added user feedback(log?, "does not support") when bias data is nullptr, it seems necessary to test this optional bias. or it would be good if test cases were added in advance for future support.

please tell me if there is something I did not understand.

**Comment by ys44kim** [comment]
@jyoungyun 
i checked the commit history and review comment again and there was something I didn't understand well.
i will check more about the cpu kernel and review whether additional improvements are needed.
thank you for your kind reply.

**Comment by jyoungyun** [comment]
@ys44kim 

> In onert the DepthwiseConv2D always has bias data (please tell me if this is wrong.)

Yes, you're right.

> and, In fact, I don't understand why it is tested when bias is nullptr because onert does not support optional bias.

The codes under `compute/cker` directory is implemented using cpu for each operation.
ONERT uses various kinds of backends. Currently, ONERT takes cpu, acl_cl, neon, trix(npu) as backend. The behavior of each backend may or may not match ONERT.

In this case, ONERT does not support optional bias_data. However, CPU kernel that I imlemented on this PR supports optional bias_data because I think it is better to support more diverse functions in the cpu kernel. If ONERT supports optional bias in the future, we can support this kernel without modifying it.

Therefore, this cpu kernel has decided to support optional bias even though ONERT does not support it. :)

**Review by ragmani — APPROVED** [review]
LGTM

**Review by ys44kim — APPROVED** [review]
LGTM

**Comment by jyoungyun** [comment]
@Samsung/one_onert  Please merge this PR. :)
